### PR TITLE
DOPE-209: added DeleteClause

### PR DESCRIPTION
**Pull request by jansigi**

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt [hunk 1]
```diff
@@ -20,4 +22,6 @@ class QueryBuilder {
     fun selectRaw(expression: SingleExpression): SelectRawClause = SelectRawClause(expression)
 
     fun selectFrom(fromable: Fromable): FromClause = SelectClause(AsteriskExpression()).from(fromable)
+
+    fun deleteFrom(from: Bucket): DeleteClause = DeleteClause(from)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:26]: name parameter just bucket

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/DeleteClause.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package ch.ergon.dope.resolvable.clause
+
+import ch.ergon.dope.resolvable.fromable.Bucket
+
+class DeleteClause(private val from: Bucket) : IDeleteClause {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/DeleteClause.kt:5]: name parameter bucket

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import ch.ergon.dope.resolvable.formatToQueryString
+import ch.ergon.dope.validtype.NumberType
+
+private const val LIMIT = "LIMIT"
+
+class LimitClause(private val numberExpression: TypeExpression<NumberType>, private val parentClause: IOrderByClause) : ILimitClause {
+    override fun toQueryString(): String = formatToQueryString(parentClause, LIMIT, numberExpression)
+}
+
+class DeleteLimitClause(private val numberExpression: TypeExpression<NumberType>, private val parent: IDeleteWhereClause) :
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt:13]: parentClause

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/OffsetClause.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import ch.ergon.dope.resolvable.formatToQueryString
+import ch.ergon.dope.validtype.NumberType
+
+private const val OFFSET = "OFFSET"
+
+class OffsetClause(private val numberExpression: TypeExpression<NumberType>, private val parentClause: ILimitClause) : IOffsetClause {
+    override fun toQueryString(): String = formatToQueryString(parentClause, OFFSET, numberExpression)
+}
+
+class DeleteOffsetClause(private val numberExpression: TypeExpression<NumberType>, private val parent: IDeleteLimitClass) :
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/OffsetClause.kt:13]: parentClause

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/ReturningClause.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package ch.ergon.dope.resolvable.clause
+
+import ch.ergon.dope.resolvable.expression.unaliased.type.Field
+import ch.ergon.dope.resolvable.formatToQueryString
+import ch.ergon.dope.validtype.ValidType
+
+class ReturningClause(
+    private val field: Field<out ValidType>,
+    private vararg val fields: Field<out ValidType>,
+    private val parent: IDeleteOffsetClass,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/ReturningClause.kt:10]: parentClause

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package ch.ergon.dope.resolvable.clause
+
+import ch.ergon.dope.resolvable.expression.TypeExpression
+import ch.ergon.dope.resolvable.formatToQueryString
+import ch.ergon.dope.validtype.NumberType
+
+private const val LIMIT = "LIMIT"
+
+class LimitClause(private val numberExpression: TypeExpression<NumberType>, private val parentClause: IOrderByClause) : ILimitClause {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt:9]: If you have a _DeleleLimitClause_ then I would name this _SelectLimitClause_, the same with the other classes and interfaces

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt (hunk 6 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt:9]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/ReturningClause.kt (hunk 5 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/ReturningClause.kt:10]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/OffsetClause.kt (hunk 4 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/OffsetClause.kt:13]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt (hunk 3 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/LimitClause.kt:13]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/DeleteClause.kt (hunk 2 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/DeleteClause.kt:5]: done

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt (hunk 1 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:26]: done

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:1]: Can you move all the Clause-files except for the Interfaces and the Clause-document into a subpackage. Maybe call it 'model'

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/IDeleteClause.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+interface IDeleteOffsetClass : IReturningClause {
+    fun returning(field: Field<out ValidType>, vararg fields: Field<out ValidType>) = ReturningClause(field, *fields, parentClause = this)
+}
+
+interface IDeleteLimitClass : IDeleteOffsetClass {
+    fun offset(numberExpression: TypeExpression<NumberType>) = DeleteOffsetClause(numberExpression, this)
+}
+
+interface IDeleteWhereClause : IDeleteLimitClass {
+    fun limit(numberExpression: TypeExpression<NumberType>) = DeleteLimitClause(numberExpression, this)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/IDeleteClause.kt:20]: here is limit-method missing (see selectWhereClause)

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:1]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/IDeleteClause.kt (hunk 7 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/IDeleteClause.kt:20]: done

**Review by pgruntz — APPROVED** [review]

**Pull request merged**